include/felspar/parse/extract.native.hpp

```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+
+#include <felspar/exceptions.hpp>
+#include <felspar/parse/concepts.hpp>
+#include <felspar/parse/extract.detail.hpp>
+
+
+namespace felspar::parse::binary::native {
+
+
+    /// ## Extract a value without checking that the span is large enough
+    template<concepts::numeric T>
+    inline T unchecked_extract(std::span<std::byte const, sizeof(T)> const s) noexcept {
+        if constexpr (sizeof(T) == 1u) {
+            return static_cast<T>(s[0]);
+        } else {
+            return detail::native_extract<T>(s);
+        }
+    }
+
+
+    template<concepts::numeric T>
+    inline T unchecked_extract(std::span<std::uint8_t const, sizeof(T)> const s) noexcept {
+        return unchecked_extract<T>(std::as_bytes(s));
+    }
+    template<concepts::numeric T>
+    inline T unchecked_extract(std::span<char const, sizeof(T)> const s) noexcept {
+        return unchecked_extract<T>(std::as_bytes(s));
+    }
+
+
+    /// ## Extract a numeric from a buffer
+    template<concepts::numeric T>
+    inline T extract(std::span<std::byte const> &s,
+                    felspar::source_location const &loc =
+                            felspar::source_location::current()) {
+        if (s.size() >= sizeof(T)) {
+            auto const v = unchecked_extract<T>(std::span<std::byte const, sizeof(T)>{s.data(), sizeof(T)});
+            s = s.subspan(sizeof(T));
+            return v;
+        } else {
+            throw felspar::stdexcept::logic_error{
+                    "Buffer doesn't have enough data", loc};
```

Comment:
Might be helpful to give a bit more context for debugging here: what were you trying to extract, how many bytes did you need, how many bytes did you get?

(There's no guaranteed way to determine a human-readable name for a type, but gcc and clang environments have `__cxa_demangle`, and Visual Studio likes human-readable mangling.  And if we're talking about a very limited set of types, there's always template variable specialisation.)

Reply:
In practice it's not very useful as the result is generally that the buffer has zero bytes in it and any error happened much earlier, but I've added the sizes and abstracted out the actual size check as well.